### PR TITLE
output-json: Enrich eve-log with geoip information from libmaxminddb

### PR DESCRIPTION
**Pull request by fandigunawan**
I faced a very large log to be enriched using geoip. At the moment I utilize logstash to enrich the log file from Suricata but at the cost of additional resources in RAM, CPU, and storage. I would like to optimize this by adding optional geoip enrichment into Suricata.

- [ X ] I have read the contributing guide lines at
   https://docs.suricata.io/en/latest/devguide/contributing/contribution-process.html
- [ X ] I have signed the Open Information Security Foundation contribution agreement at
   https://suricata.io/about/contribution-agreement/ (note: this is only required once)
- [ X ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: none

Describe changes:
- I adds optional geoip enrichment into Eve log by setting geoip-city-database option under eve-log configuration in suricata.yaml
- The JSON structure of geoip is based on [Elastic ECS geo specification](https://www.elastic.co/guide/en/ecs/current/ecs-geo.html) 

Eve output example with enrichment
```
{
    "timestamp": "2021-05-27T03:37:44.575843+0700",
    "flow_id": 2130735805455113,
    "pcap_cnt": 15756,
    "event_type": "fileinfo",
    "geoip_src": {
        "ip": "192.236.155.230",
        "geo": {
            "continent_code": "NA",
            "country_iso_code": "US",
            "city_name": "Seattle",
            "country_name": "United States",
            "continent_name": "North America",
            "timezone": "America/Los_Angeles",
            "location": {
                "lat": 47.4902,
                "lon": -122.3004
            }
        }
    },
    "geoip_dst": {},
    "src_ip": "192.236.155.230",
    "src_port": 80,
    "dest_ip": "10.5.26.4",
    "dest_port": 56042,
    "proto": "TCP",
    "pkt_src": "wire/pcap",
    "http": {
        "hostname": "192.236.155.230",
        "url": "/images/redbutton.png",
        "http_user_agent": "WinHTTP loader/1.0",
        "http_content_type": "Content-type: application/octet-stream",
        "http_method": "GET",
        "protocol": "HTTP/1.1",
        "status": 200,
        "length": 105556
    },
    "app_proto": "http",
    "fileinfo": {
        "filename": "/images/redbutton.png",
        "gaps": false,
        "state": "TRUNCATED",
        "stored": false,
        "size": 102400,
        "tx_id": 0
    },
}
```

### CODE REVIEW

**Comment by jlucovsky** [comment]
Thanks for your contribution.

I'd like to see the *inclusion of GeoIP data* into the output stream configurable.

Can the new configuration `geoip-city-database` be shared in any way with the existing `geoip-database` configuration setting?

File: src/output-json.c [hunk 1]
```diff
@@ -77,6 +77,86 @@
...
+static MMDB_s mmdb;
+static int mmdb_status = MMDB_FILE_OPEN_ERROR;
+
+void GeoIPGet(JsonBuilder *js, const MMDB_s *const mmdb, const char *ip_address, const char *key);
+
+inline void GeoIPSetString(JsonBuilder *js, MMDB_entry_data_s *entry_data, const char *key)
+{
+    char *str = NULL;
+    if (entry_data->has_data && entry_data->utf8_string != NULL) {
+        str = SCStrndup((char *)entry_data->utf8_string, entry_data->data_size);
```
Comment (jlucovsky) [inline src/output-json.c:92]: This isn't needed. The output is staged into a buffer into which the string will be copied.

**Comment by jasonish** [comment]
First commit subject needs a cleanup.

File: doc/userguide/output/eve/eve-json-output.rst [hunk 2]
```diff
@@ -27,6 +27,10 @@ Output types::
 
       filetype: regular #regular|syslog|unix_dgram|unix_stream|redis
       filename: eve.json
+      # Enable GeoIP on source ip and destination ip
+      # Enter full path of GeoLite City database to enrich eve-log with geoip
+      # Disable geoip enrichment by commenting geoip-city-database
+      #geoip-city-database: /usr/share/GeoIP/GeoLite2-City.mmdb
```
Comment (jasonish) [inline doc/userguide/output/eve/eve-json-output.rst:33]: We already have a "geoip-database" configuration section, is there a reason we need another?  AFAIK, there other can be a city or a country database, depending on what you have access to.

**Comment by victorjulien** [comment]
Needs to target `master`, then we can consider backporting later.

**Comment by victorjulien** [comment]
Fundamentally: is it the task of Suricata to do this enrichment, or should it be done in post processing?

**Comment by victorjulien** [comment]
Current `geoip` support includes detection, would like to hear your ideas on adding detection support here as well.

**Comment by jasonish** [comment]
> Fundamentally: is it the task of Suricata to do this enrichment, or should it be done in post processing?

If we can resolve the logging and detection databases to be the same, its a source of truth in the event for the detection.

And to get really picky, even the rule message could be left to post-processing.  But I think its that source of trust above that might make it worth it.

File: doc/userguide/output/eve/eve-json-output.rst (hunk 2 above)
Comment (fandigunawan) [reply doc/userguide/output/eve/eve-json-output.rst:33]: Yes, you are right. Perhaps, it is better to merge it with "geoip-database". I was initially thinking that I can save few MBs of memory if this feature is optional due to the implementation of libmaxmindb that use mmap.

If we separate City vs Country we could save memory if enrichment is optional.
For instance these are the size of the databases:
-rw-r--r--.   1 root root  60M Dec 18  2019 GeoLite2-City.mmdb
-rw-r--r--.   1 root root 3.9M Dec 18  2019 GeoLite2-Country.mmdb

**Comment by fandigunawan** [comment]
> Needs to target `master`, then we can consider backporting later.

I will close the PR and target master instead.

File: doc/userguide/output/eve/eve-json-output.rst (hunk 2 above)
Comment (jasonish) [reply doc/userguide/output/eve/eve-json-output.rst:33]: Would there be a reason why one would want to use city for one, and country for another?

Comment (fandigunawan) [reply doc/userguide/output/eve/eve-json-output.rst:33]: Okay, I will revise the code.

**Comment by fandigunawan** [comment]
> Some changes were requested (like targeting `master` branch)

I made second PR in https://github.com/OISF/suricata/pull/10703 .

**Pull request closed without merging**